### DISCUSSION TITLE
How can I check the empty lines between variables?

### DISCUSSION THREAD

**Discussion by gabryellr** [checkstyle/checkstyle#14605] [Q&A] [answered]
I'm trying to validate if the class has the annotated variable declared with empty line between them, like this:

```
    @Value("")
    private int a;

    @Value("")
    private int b;
```
I didnt found any rule that validate it, so I started to look at checkstyle GUI but I cannot see empty lines on GUI nor CMD. How can I achieve it? thanks

![image](https://github.com/checkstyle/checkstyle/assets/30318979/af3f7d1f-7910-4dec-8299-f8e22b64ad37)

**Comment by nrmancuso** [comment]
I would suggest that you try out  [EmptyLineSeparator](https://checkstyle.sourceforge.io/checks/whitespace/emptylineseparator.html#EmptyLineSeparator) for this.

> Checks for empty line separators before package, all import declarations, fields, constructors, methods, nested classes, static initializers and instance initializers.

**Reply by gabryellr** [reply]
the problem is I can do it too:

```
    private int a;
    
    private int b;
```

I would like to force a empty line between variables that are annotated but keep the variables with no annotation together. The final result should be:

```
    @Value("")
    private int a;
    
    @Value("")
    private int b;

    private int a;
    private int b;
```

**Reply by nrmancuso** [reply]
You can create an XPathSuppression to avoid placing violations on fields that do not have annotations.

**Reply by romani** [reply]
>but I cannot see empty lines on GUI nor CMD

Empty lines are not part of AST. But fact that you have no nodes on line 2 and 5 imply meaning that there are empty lines.

You need to write custom Check to validate what you want.

**Reply by gabryellr** [reply]
can you give some examples checking no nodes? I tried using the MatchXpath but its a bit trick to understand it

**Reply by nrmancuso** [reply, marked as answer]
> You need to write custom Check to validate what you want.

Unless I am missing something, this is overkill.

>MatchXpath

MatchXpath is going to be very difficult or impossible to use to achieve this. I suggest that you go with my suggestion above: use EmptyLineSeparator with a SuppressionXpathSingleFilter.

Example:
```
➜  src cat Test.java

public class Test {
    @Value("")
    private int a;

    @Value("")
    private int b;

    private int c;
    private int d; // ok, no annotation

    @Value("")
    private int e;
    @Value("")
    private int f;// violation, no line break between fields with annotation

    void m() {
        int x = 2;
        int y = 3; // ok, not a field
    }

    @interface Value {
        String value();
    }
}

➜  src cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
  <module name="TreeWalker">
    <module name="SuppressionXpathSingleFilter">
      <property name="checks" value="EmptyLineSeparator"/>
      <property name="query"
                value="//CLASS_DEF/OBJBLOCK/VARIABLE_DEF[not(./MODIFIERS/ANNOTATION)]"/>
    </module>
    <module name="EmptyLineSeparator">
      <property name="tokens" value="VARIABLE_DEF"/>
    </module>
  </module>
</module>

➜  src java -jar checkstyle-10.13.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:14:5: 'VARIABLE_DEF' should be separated from previous line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 1 errors.

```

You can fine tune this with exact annotation names or name pattern if you need to.

**Reply by gabryellr** [reply]
Thank you, I'll take a look on it

**Reply by gabryellr** [reply]
@nrmancuso thanks for the replies but for example in these cases the rule is not being applied:

```
    @Value
    int a; //should have a empty line after the variable a
    int b;

    int c; 
           //this empty line should not exist
    int d;
```

sorry, I'm trying to understand it

**Reply by nrmancuso** [reply]
Ok, it sounds like you have some more specific requirements than your original post specified. For:
```
    @Value
    int a; //should have a empty line after the variable a
    int b;
```

you can just tweak the XPath suppression to something like `//CLASS_DEF/OBJBLOCK/VARIABLE_DEF[not(./MODIFIERS/ANNOTATION)
                 and not(./preceding-sibling::*[1]/MODIFIERS/ANNOTATION) ]"`.

For:
```
    int c; 
           //this empty line should not exist
    int d;

```
you would need to write your own check.
